### PR TITLE
Add a cache for previously transformed code

### PR DESCRIPTION
**Pull request by mauricedb**
Added a cache to prevent transforming the same code every time a Gulp watch triggers a new build. As usually most ES6 files are not actually changed there is no need to transform the again which could be costly.

In our build this change resulted in a drop from 15 seconds to 3 seconds to Browserify and Babelify our React code.

### CODE REVIEW

**Comment by jmm** [comment]
@mauricedb Can you summarize why adding caching behavior here is preferable to using something like watchify?

**Comment by zertosh** [comment]
@mauricedb Like @jmm pointed out, can you clarify why you'd use this over https://github.com/substack/watchify ?

**Comment by mauricedb** [comment]
@jmm and @zertosh,

Watchify works great if you know about it and manage to configure it properly. Most people, myself included until quite recently, don't know about it and quite frankly when I first came across it I was under the impression that it was a replacement for gulp.watch(). My bad :-(

With this PR you get most of the speed benefits of watchify automatically for free. No need to do or configure anything. It will prevent calling babel for the same file contents and just return the previous result. This solves most peoples performance problems with gulp tasks automatically.

So it doesn't replace watchify it just makes the need much smaller.

**Comment by zertosh** [comment]
Sorry @mauricedb but I'm a "no" on this, for several reasons:

1. It only caches babelify results, which means that browserify still has to tons of other work (including dependency resolution). watchify handles all of that.

2. Your cache assumes that `babel.transform` is a pure function - it's not. The output not only depends on source code, but on the options passed in, and the babelrc file. If you're running multiple builds, which diff configs, you're gonna break one of them.

Thank you for the contribution though. If you need any help with watchify, either I or @jmm would be more than happy to help out.

**Comment by mauricedb** [comment]
I have watchify working for my normal bundle based on [this](https://github.com/gulpjs/gulp/blob/master/docs/recipes/fast-browserify-builds-with-watchify.md) recipe. Maybe adding a link to the FAQ sheet would be helpful for others.

The two things I haven't figured out yet is how to speed up the preprocessing for Karma and for eslint. For eslint I am using gulp-cached so far to only lint the changed files but that means that I don't see errors in files I did not just edit. Not ideal but a workable solution.

Any pointers would be appreciated.

Thanks.

**Comment by zertosh** [comment]
@mauricedb I can't help you with Karma. As for eslint, why are you running the post-processed files through it, instead of the pre-processed ones?

**Comment by mauricedb** [comment]
@zertosh We aren't pre-processing files before running eslint. Instead we are using babel-eslint which uses babel-core (not babelify). The reason of using babel-eslint over plain eslint is that we are using React and plain eslint doesn't handle the JSX code while babel-eslint does.

The end result is that both karma and babel-eslint still take quite a bit of time. These are less important though as the babelify action is fast now with watchify and that lets developers stay in the flow.
